File: 05 Data Structure/Data Structure - split_array_to_k_subarray_find_min_of_max_sum_of_each_combination.cpp
// Approch: Binary Search
/*
Split the given array into K sub-arrays such that maximum sum of all sub arrays is minimum.
*/

#include <bits/stdc++.h>
using namespace std;
 
bool check(int mid, int array[], int n, int K){
    int count = 0;
    int sum = 0;
    for (int i = 0; i < n; i++) {
        if (array[i] > mid)
            return false;
        sum += array[i];
        if (sum > mid) {
            count++;
            sum = array[i];
        }
    }
    count++;
    if (count <= K)
        return true;
    return false;
}

int solve(int array[], int n, int K){
    int* max = max_element(array, array + n);
    int start = *max;
    int end = 0;
 
    for (int i = 0; i < n; i++) {
        end += array[i];
    }
    int answer = 0;
    while (start <= end) {
        int mid = (start + end) / 2;
        if (check(mid, array, n, K)) {
            answer = mid;
            end = mid - 1;
        }
        else {
            start = mid + 1;
        }
    }
 
    return answer;
}

int main(){
    int array[] = { 1, 2, 3, 4 };
    int n = sizeof(array) / sizeof(array[0]);
    int K = 3;
    cout << solve(array, n, K);
}

// Note: From https://www.geeksforgeeks.org/split-the-given-array-into-k-sub-arrays-such-that-maximum-sum-of-all-sub-arrays-is-minimum/

File: 05 Data Structure/Data Structure - split_array_to_k_subarray_find_min_of_max_sum_of_each_combination.py
# Approch: Binary Search
'''
Split the given array into K sub-arrays such that maximum sum of all sub arrays is minimum.
'''
# Version 1:
def check(mid, array, n, K):
    count = 0
    sum = 0
    for i in range(n):
        if array[i] > mid: return False
        sum += array[i]
        if (sum > mid):
            count += 1
            sum = array[i]
    count += 1
    if count <= K: return True
    return False

def solve(array, n, K):
    start = max(array)
    end = 0
    for i in range(n):
        end += array[i]
    answer = 0
    while start <= end:
        mid = (start + end) // 2
        if (check(mid, array, n, K)):
            answer = mid
            end = mid - 1
        else:
            start = mid + 1
    return answer

N, K = list(map(int, input().split()))
l = list(map(int, input().split()))

print(solve(l, N, K))
#######################################
# Version 2: Antt (Anthoy Sheu)
def solve():
    s, e = max(f), sum(f)
    ans = 0
    while s <= e:
        mid = (s+e)//2
        if check(mid): ans = mid; e = mid-1
        else: s = mid+1
    return ans

def check(mid):
    c, s = 0, 0
    for i in range(n):
        if f[i] > mid: return 0
        s += f[i]
        if s > mid: c += 1; s = f[i]
    c += 1
    return c <= m

n, m = list(map(int, input().split()))
f = list(map(int, input().split()))

print(solve())

# Note: From https://www.geeksforgeeks.org/split-the-given-array-into-k-sub-arrays-such-that-maximum-sum-of-all-sub-arrays-is-minimum/
